### DISCUSSION TITLE
Question about ST7789 display driver functionality

### DISCUSSION THREAD

**Discussion by Sum-Guy** [arendst/Tasmota#17513] [Q&A] [answered]
My question pertains to the ESP8266.  If the situation on the ESP32 is different, please let me know.

There are some ST7789-based display screens that have (or don't have) I/O pins labelled as CS and/or RES.  I don't know if there are boards that have both.

It seems that the Tasmota driver for the ST7789 anticipates the presence of the CS control line, but not the RES control line.  From what little I can find, the RES control line is a RESET line and "this signal will reset the device and it must be applied to properly initialize the chip.  Signal is active low."  I take it that this line must go low and then high for the display to be initialized.  If this needs to happen periodically depending on what you're doing with the display, I don't know.

The CS (Chip Select) line seems to have a different functionality and purpose. This line is active low.

Can I get some verification that Tasmota with display drivers will, or won't, correctly work with an ST7789 board with a RES control line?

Bonus question:  Tasmota seems to also want to define the MISO signal line when working with SPI in general or the ST7789 in particular, but there is nothing to connect that line to.  I assume it's not needed in this case?

**Comment by sfromis** [comment]
To define a reset control line for a SPI display, you can config a `Display Rst` pin, I don't know if/how that might affect what is done for such a ST7789 display (I've only got it with parallel interface instead), but I'd do that config.

I dont know what this means, but in general CS has a very specific purpose for SPI, to let the device know that it has the bus available.
> The CS (Chip Select) line seems to have a different functionality and purpose.

I think that MISO is only relevant when the SPI device can send something back to the controller, and that is not a given for non-touch displays.

More generally, I'd look in the direction of using the newer uDisplay driver instead of the old chip-specific drivers, but I can't give a real pro et contra there.

**Reply by Sum-Guy** [reply]
These ST7789 display modules (240 x 240) do not bring the CS line out to a pin.  They tie that line low.  They bring the RES (reset) line out.  In tasmota you have ST7789 CS and ST7789 DC.  DC is of course required, but since the CS pin is not available, is the tasmota driver ok if you don't define the CS line?   

And get this:  I issue the command "displaymode 1" and I get nothing on the display.  I issue the command "displaymode 4" and the esp8266 crashes and re-boots!  (I have a DHT11 connected and I do get data from it).

Even though I do not define this in the template, tasmota automatically assigns D6 GPIO12 to  SPI MISO,  But I have nothing to connect that to. 

I've assigned the display RES pin to tasmota Display rst, but that doesn't seem to help.

Side question:  the commands displaywidth and displayheight both come back with 240, which is correct for this display module, but I didn't define those numbers.  Does tasmota assume those display dimensions for the ST7789 or does it query the display to get them?

On this link:  https://tasmota.github.io/docs/Commands/#displays
Why are displaymodels 11 through 14 not listed?  I note that the ST7789 should show up as #12 on that list.

I've flashed the setup I have with an arduino sketch that prints stuff on the display, and it does work.  So I know the display and the lolin nodemcu are working and the wiring connection setup is ok.  I really would like to know if anyone has gotten these displays to work with Tasmota, and if they did, how was it configured?

**Reply by sfromis** [reply]
First, a common tripwire is that display drivers commonly add a power channel for powering on/off the display, via Tasmota front page toggle or Power command. It happens that people overlook this :smile: 

If no CS line is available, that should mean that the display assumes full control over the SPI bus, which does not explain your "have a different functionality and purpose". My guess would be that the driver can manage without this, but you could try just defining it on some unused pin...

More generally, you did not tell your specific pin config, which should use the hardware SPI pins (gpio 12=MISO, 13=MOSI, 14=CLK, 15=CS), without doing anything else with those. If your display does not provide all of those, it should be able to manage without.

**Reply by Sum-Guy** [reply]
Under tasmota github docs Displays I see, under the Displaytext command, these miscellaneous commands:

o = switch display off
O = switch display on

How do I use them?

Is it "displaytext O" to switch the display on?

There is also a "dimmer" command (displaydimmer).  Is it relavent here?

The following is also from the github Display docs:

The ST7789 is connected via 4 Wire software SPI ((ST7789_CS), SSPI_SCLK, SSPI_MOSI, ST7789_DC, OLEDRESET, Backlight ) 

Should I be seeing an "OLEDRESET" option in the module config list?  And use it instead of DisplayRST?  Because I'm not seeing it.  I note that "ST7789_CS" is in brackets ( ) which means that it's optional?

I want to know:  If the ONLY display-related command I issue is "displaymode 1" or "displaymode 2", is that sufficient to get tasmota to display *something* without having to issue more commands?

Regarding wire hookup, this is taken from the output during boot up:

 SPI: Hardware using GPIO14(CLK), GPIO13(MOSI) and GPIO12(MISO)
 DSP: ST7789

DHT11 does not show up here, should it?  (It is connected, I do see data on the main web gui page).


In module config I have:

D3 GPIO0 - ST7789 DC (to display DC)
D2 GPIO4 - ST7789 CS (yes I've defined it, no connection)
D1 GPIO5 - DHT11 (to DHT11)
D6 GPIO12- SPI MISO (no connection)
D5 GPIO14- SPI CLK (to display SCK)
D7 GPIO13- SPI MOSI (to display SDA)
D0 GPIO16- Display Rst (to display RES)

Regarding the ESP crashing when the displaymode 4 command is issued - I don't have mqtt configured but that command does involve mqtt, is that the reason for the crash?

Here's the output from the display command:

CMD: display
RSL: RESULT = {"Display":{"Model":12,"Type":0,"Width":240,"Height":240,"Mode":2,"Dimmer":10,"Size":1,"Font":1,"Rotate":0,"Invert":0,"Refresh":2,"Cols":[16,8],"Rows":2}}

So my questions now are:

1) if a "display power" command is needed, what exactly is the command?

2) are the commands "displaymode 1" or "displaymode 2" sufficient by themselves to get something to be printed on the display? 

3) There is mention of an OLEDRESET parameter in module config but I'm not seeing it.

**Reply by sfromis** [reply]
No, to switch on, use the front page toggle or the power command, just like for a relay or light. Has nothing to do with display support. This assumes that you got a power toggle on the front page when having the display configured and recognized, which seems to be the case based on the messages you posted. Whatever you have of display commands alluding to power or whatever is unrelated to the very basic thing I pointed out.

"OLED Reset" would probably be the `Display Rst` I already mentioned, not instead of. Don't know if CS makes a difference, but I specifically suggested to use the hardware SPI pins, which would be gpio 15 for `ST7789 CS`, not any other.

**Reply by Sum-Guy** [reply]
First of all, I can find nothing on the Configure Module screen where the front page toggle button has been defined.   So to me its confusing what the device is doing when the front page toggle button is clicked.  I do have GPIO2 defined as Linkled, which does work, but pressing the toggle button to the "ON" state turns the linkled off, and vice versa.  I don't know where it's been defined that the toggle button function is tied to GPIO2 or the linkled.

I can confirm that a console power command returns POWER = ON or OFF depending on the front page toggle state.  But again where or how the state of that toggle button is defined or how the device is using the state of that toggle is not obvious to me.

It would not make sense for the front-page power toggle button to be linked to a connected display device.  What if I had a relay tied to the front page toggle switch?  Are you saying that when I want the relay to be on or off, that the display would also turn on or off?  That I would not have independent control of the relay AND the display?

When the device is reset, I see transiently for about a second the screen comes on and it looks like a static display of random grey noise on the screen.

And I can confirm that regardless if the POWER state is ON or OFF, it has no effect on whether or not I can get anything to show up on the display.  I've set dimmer to 100, I have the backlight tied to vcc.  I've issued the commands "displaymode 1" and "displaymode 2" but still I get nothing displayed.

I do not have ST7789_CS defined in the module configuration.  If I define ST7789_CS to gpio15, it makes no difference, still not able to get anything on the display.  I am curious as to why you say CS has to be on GPIO15.  It seems that CS and DC can be on any available GPIO. I have DC on GPIO-0.    See my above post for the entire pin hookup.

As I currently have the device configured, here is the result of the GPIO command, if this helps:

RESULT = {"GPIO0":{"6144":"ST7789 DC"},"GPIO1":{"0":"None"},"GPIO2":{"544":"LedLink"},"GPIO3":{"0":"None"},"GPIO4":{"0":"None"},"GPIO5":{"1184":"DHT11"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO12":{"672":"SPI MISO"},"GPIO13":{"704":"SPI MOSI"},"GPIO14":{"736":"SPI CLK"},"GPIO15":{"6112":"ST7789 CS"},"GPIO16":{"1024":"Display Rst"},"GPIO17":{"0":"None"}}

Here is the result of the display command:

RESULT = {"Display":{"Model":12,"Type":0,"Width":240,"Height":240,"Mode":2,"Dimmer":100,"Size":1,"Font":1,"Rotate":0,"Invert":0,"Refresh":2,"Cols":[16,8],"Rows":2}}

Are we at a dead end here, or is there anything else you would suggest I check or do?

**Reply by sfromis** [reply]
> First of all, I can find nothing on the Configure Module screen where the front page toggle button has been defined.

Right, it is just a feature of a display driver, not something you configure separately. My point was just that **if** you got a toggle and did not turn it on, don't expect any display activity. That has been "surprising" to more than a couple of users. And it does very much make sense independently of your musings; if you also have two relays, you just gets two additional toggles. And if you have, say, a WS2812 strip, you get yet another toggle and some sliders.

Whatever happens on your LedLink is just a status indication of connection state or trouble, not related to displays.

When I suggest CS on gpio15 it is simply because hardware SPI pins on ESP8266 is gpio 12-15, as already explained, therefore it makes sense to use these 4 pins exactly for the hardware SPI meaning.

Anyway, I've not used much in old driver displays on ESP8266, strongly preferring ESP32, uDisplay, and LVGL/HASPmota.

**Reply by Sum-Guy** [reply]
If you think that CS should be on GPIO 12 - 15 then does that also apply to DC (ST7789_DC) ?

**Reply by sfromis** [reply]
No, I did not say that CS should just be on one of gpio 12-15, but exactly on gpio 15.

As I already described:
> should use the hardware SPI pins (gpio 12=MISO, 13=MOSI, 14=CLK, 15=CS), without doing anything else with those. 

DC is not among the hardware SPI pins, and thus no fixed placement.

**Reply by Sum-Guy** [reply]
Well, this is now a dead end.

I have done everything you have suggested, I have configured all pins as suggested.  Yet Tasmota is not displaying anything on this ST7789-based display device.

Unless someone reading this has first hand experience and has seen Tasmota work correctly with these particular display modules, I must conclude that tasmota's driver implimentation for this display module is not functional.

**Reply by sfromis** [reply]
What I tried to do was only to point out some basic details, not assuming to know what more it may take to make it work, and would not be surprised if I just did not know enough.

**Reply by Sum-Guy** [reply]
These ST7789 240 x 240 displays are incredibly cheap and seem to look good.  I don't know if there are any other display modules that have a similar or higher pixel-count for under $10.

Is there nobody else here that has any direct experience with them and tasmota?

**Reply by Sum-Guy** [reply]
I've read through this:

https://github.com/gemu2015/Sonoff-Tasmota/issues/28#issuecomment-751881775

I don't know if it contains anything relevent here.  It applies to the ESP32.

Also, regarding having to use the toggle switch (power button) set to ON for the display to work, that may be because the backlight control like might be connected to a GPIO pin, and if so defined in tasmota then tasmota will assign the power toggle to control the backlight.

This thread looks interesting:

https://github.com/arendst/Tasmota/discussions/13161

Is it possible to send a message to user [gemu2015] ?  I can't figure out how to contact him directly.

**Reply by sfromis** [reply]
I can just observe that the Tasmota front page power toggle added for a display is important for I2C displays without any separate backlight pin, implying that at least that display driver uses the toggle actively.

If you @-mention gemu2015, chances are that it will be noticed.

**Reply by sfromis** [reply]
FWIW, as a reference point to verify my understanding about Tasmota old driver support for SPI displays on ESP8266, I set up a test rig with ESP8266 and ILI9341, using the `tasmota-display` binary, and configured pins for the purpose:
```
Gpio.00     D3    1 User
Gpio.01     TX    1 User
Gpio.02     D4  992 Backlight
Gpio.03     RX    1 User
Gpio.04     D2 5920 ILI9341 DC
Gpio.05     D1 1024 Display Rst
Gpio.09     FL    0 None
Gpio.10     FL    0 None
Gpio.12     D6  672 SPI MISO
Gpio.13     D7  704 SPI MOSI
Gpio.14     D5  736 SPI CLK
Gpio.15     D8 5888 ILI9341 CS
Gpio.16     D0    1 User
Gpio.17     A0    1 User
```
Booting with this, powering the display up using the toogle, and doing a `DisplayMode 1` to have Tasmota generate some display activity, it worked right away.

**Reply by sfromis** [reply]
BTW.... When it comes to cheap display modules with decent resolution, the classic ILI9341 320x240 is worth considering, also under $10.

**Reply by Sum-Guy** [reply]
I'll read and digest your last post a little later - just wanted to post the following:

On a fresh flash of tasmota with sensors and display, with nothing defined in the module config, there is no toggle button on the main page.

The minute you define the pins for the ST7789 and save, then bring up the main page, the toggle button and state are displayed.  So yea I guess there is something to this toggle power turning the display on and off.  But what is it doing?  

Is it controlling the CS, or the back-light?  Since I tie the backlight to VCC, and since the display has CS tied low on the board, the ESP8266 really can't turn the display on or off.  

Note that I have yet to define backlight in the module config.  Does tasmota have a problem with that?

Another random observation:  On the page https://tasmota.github.io/docs/Displays/ there is no mention of the command DisplaySize.   But it is mentioned on this page:  https://tasmota.github.io/docs/Commands/#displays

Random question - what is "displayaddress" ?  

CMD: displayaddress
RSL: RESULT = {"DisplayAddress1":113}

Am I supposed to use that address somehow?  In any display text commands or settings?

Another random question:   Displaytype - what is that?  How is it related to displaymodel?

This: https://tasmota.github.io/docs/Commands/

says DisplayType - Select display sub-modules. More info...

More info is linked to https://tasmota.github.io/docs/Displays/

But there is no mention of displaytype there.  Just Displaymodel.  My display model is set to 12, displaytype is set to 0.

**Reply by sfromis** [reply]
Great that you got to a working setup, makes it much easier to experiment with your hammering about what you need or not.

`DisplayAddress` is for I2C displays like SSD1306, not SPI displays.

`DisplayType` is an option relevant to ILI9341 (never used it) and TM16xx.

The only reset I've seen for displays is the `Display Rst` config I've mentioned already. Not "replaced", but sometimes different names for same thing, don't know the backstory.

**Comment by Sum-Guy** [comment]
I went back and flashed tasmota-display.bin verion 9.4.0 (this is for esp8266) 

**AND IT WORKS.**

I can put text on the screen (displaytext what-ever).   Toggling the front-page power switch does turn the display on and off.  What-ever text I had on the screen will come back when I go from off to on.  

I note that the parameter "Display RST" is abscent, seemingly replaced by "OLED Reset".  I have that on GPIO4.

I thought at first that this older firmware did not work, but then I defined ST7789_CS (which technically I don't have) to be on GPIO15.  I'm not sure if doing that is what is making this particular setup work, I'll have to experiment.    I have not defined Backlight.

displaymode 1 (display date and time) does nothing.  gives me a blank screen.  Using the command displaytext in this mode does nothing.

But displaymode 0 then using displaytext does put text on the screen, and displaysize does increase text size.

Ok, the CS line does not have to be defined in the configuration for displaytext to work.  

Now, why doesn't displaymode 1 work?  It just clears the screen.  The device is contacting ntp server, it has the correct date and time.

**Reply by sfromis** [reply]
Is there any point in choosing to flash a badly outdated version, instead of current?

**Reply by Sum-Guy** [reply]
What do you mean - is there any point?

Of course there's a point - the point is that AT LEAST IT WORKS.  If this is the only way that tasmota will print to the ST7789 display, then what other choice is there?  What do I lose if I use 9.4.0 vs 12?

**Reply by sfromis** [reply]
Are you saying that it only works after digging for outdated stuff?

Asking for the point was mostly that you seemed to do an extra effort to go back in time, when current version should be the first choice. And current Tasmota does have a lot of enhancements, but obviously I cannot know what difference it may make to you.

One aspect is that if you read docs for current Tasmota, you'll find features not available in older version. That alone is a great reason to use current Tasmota, unless you have a very specific gain by going back in time. A point....

**Reply by Sum-Guy** [reply]
What do I have to do to make the ST7789 work with the current version (12) of tasmota?

I think we've established that IT DOESN'T WORK using the pre-compiled display binaries and setting up the device configuration to use the ST7789 implicitly.

@gemu2015 seems to know more about this, saying the driver is "outdated".  Is it outdated, or did something break after 9.4.0?

I need EXACT instructions on how to modify user_config_override.h and / or platformio_override.ini to compile a new tasmota-display.bin under version 12.what-ever to achieve a working ST7789 display that will work in displaymodes 0, 1 and 2 (assuming the pre-compiled tasmota-display.bin is not sufficient).

I'm not against using the universal display driver - but I see no point using it if in the end I can't get displaymodes 1 and 2 working (display date/time or sensor data). 

And not that I want to open another can of worms, I have been focused so far on the ESP8266.  Is this situation with the ST7789 any different with the ESP32?  If I want to putz with the ST7789 and tasmota, should I move to the ESP32?

**Reply by sfromis** [reply]
It is not very clear what your reasoning with the old Tasmota version is. As I read what you posted, you got the ST7789 to work using the precompiled `tasmota-display` in the old 9.4 version. Correct? That was my basis for wondering if you found that the same did not work using the current version instead, which could be a plausible point in digging old stuff up.

When gemu2015 suggests to use the universal display driver instead of the old drivers, it is the same as when I mentioned uDisplay (shorter name for the same) as the preferred strategy, which you seemed to ignore. 

Universal Display (uDisplay) is a newer and improved display driver architecture, which does not need separate C++ code for every display type, but can use a common driver, based on a display descriptor to define the characteristics of each display. Thus, new display types can be supported this way, a short text file, with the possible exception of a display needing methods not previously available, like when support for parallel displays were introduced.

The word "outdated" is about the old display driver concept, not a question of 9.4 vs current. And gemu2015 specifically told you that display modes 1-5 are not supported for ST7789 with the old driver, but is supported with the universal display driver. This seems to be something you're wanting to have.

When I prefer ESP32 for such displays, much of the reasoning is greatly enhanced ways of programming for the displays using LVGL and especially HASPmota. These features depend on the Berry language, which became available with Tasmota only on ESP32.

**Reply by Sum-Guy** [reply]
YES YES YES

The only way I got tasmota to do anything with the ST7789 was to go back and use 9.4.0 display.bin.  I thought I've been very clear about what I was doing from the very start.

And so far the only thing I can do is displaytext (displaymode 0).  Displaymode 1 and 2 does not work (not even in 9.4.0).  

I have yet to try to use the universal display driver.  I need exact instructions how to use it.  I need to know if I can use it with the pre-compiled display.bin or if I need to custom compile a new bin.

"And gemu2015 specifically told you that display modes 1-5 are not supported for ST7789 with the old driver"

I have yet to get displaymode 0 (text mode) to work using the ST7789 driver under tasmota 12.  No displaymode works using the ST7789 driver under tasmota 12.  And displaymodes 1 and 2 (haven't tried others) don't work under 9.4.0.

**Reply by gemu2015** [reply]
beging the author of both the old and the new display driver i can assure that the "old" st7789 has not changed since 9.4.0 and therefore should also work with latest tasmota.
you should select SPI_MOSI, SPI_CLK,ST7789_DC,Display_RST, and Backlight.  (no need for ST7789_CS)
this should initialize your display and you could work with display mode 0

the new universal display however has advantages and also works with display mode 1-5

you should make yourself familiar with compiling your own binary, i am not aware if there is a precompiled version with uDisplay for the ESP8266 is available.

**Reply by Sum-Guy** [reply]
The ST7789 should be removed from the Configure Module list in the web GUI if it is known that the driver does not work, if no displaymodes work for that display.  Am I wrong about that?  Or does it still need to be there to use the universal display driver?

**Reply by gemu2015** [reply]
this driver works, but does not support display modes 1-5, that's all.
you may display any value you want using rules or scripts with display mode 0

**Reply by Sum-Guy** [reply]
I'm using the benzino77 tasmocompiler on gitpod to create a sensor+display bin.

I need EXACT instructions on how to modify user_config_override.h and / or platformio_override.ini (or maybe some other h or ini file?) to use the universal display driver with ST7789 display.

But I still have this question, which is not answered:  Does (or, Did) displaymode 1 and 2 work with the "old" driver under Tasmota 9.4.0?

AND -

My experience continues to be that text displaymode 0 DOES NOT WORK using ST7789 driver under tasmota 12 display bin.

**Reply by sfromis** [reply]
> But I still have this question, which is not answered: Does (or, Did) displaymode 1 and 2 work with the "old" driver under Tasmota 9.4.0?

Again: As gemu2015 specifically told you already, no.

**Comment by gemu2015** [comment, marked as answer]
this driver is outdated. better use universal display driver. it is also supported on ESP8266 even on 1M devices.

however if you use it the following facts are:

1. the backlight if not defined in Tasmota io config is fixed to Pin 2
2. defining CS is not required 
3. defining MISO is not required
4. displaymodes 1-5 are not supported, only display mode 0 with displaytext, but supported in universal driver

**Reply by Sum-Guy** [reply]
I don't understand.

Are you saying that 9.4.0 will let me print text to the display (displaymode 0) but no other displaymode will work?

Are you saying that with 12, it is known that the ST7789 driver does not work at all (in any mode) and I have to use "universal driver"?  

Are you saying that with universal driver, it will work in all modes?
